File: my-folder/1056-capacity-to-ship-packages-within-d-days/solution.cpp
class Solution {
public:
    int check(vector<int>& v, int maxCap, int days) {
        int day = 1;
        int curCap = 0;
        
        for (int i:v) {
            if (curCap + i <= maxCap) {
                curCap += i;
            } else {
                curCap = i;
                day++;
                if (day > days || i > maxCap) {
                    return 0;
                }
            }
        }
        return 1;
    }
    
    int shipWithinDays(vector<int>& weights, int days) {
        int low = 1;
        int high = 0;
        
        for (int weight : weights) {
            high += weight;
        }
        
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (check(weights, mid, days)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        return low;
    }
};

File: my-folder/1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {
    boolean solve(int[] weights, int days,int guess){
        int count = 0;
        int current = 0;
        for(int i=0;i<weights.length;i++){
            if(current+weights[i]<=guess){
                current+=weights[i];
            }
            else{
                count++;
                current = weights[i];
                if(count>days) return false;
            }
            
        }
            if(current!=0) count++;
            return count<=days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int ans=-1;
        int left = Arrays.stream(weights).max().getAsInt();
        int right = Arrays.stream(weights).sum();
        while(left<=right){
            int mid = left + (right-left)/2;
            if(solve(weights,days,mid)){
                ans = mid;
                right = mid-1;
            }
            else left = mid+1;
        }
        return ans;
        
    }
}
